onnxruntime/core/optimizer/pre_shape_node_elimination.cc

```diff
@@ -0,0 +1,64 @@
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/pre_shape_node_elimination.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status PreShapeNodeElimination::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  if (graph_utils::RemoveNode(graph, node)) {
+    rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+  }
+
+  return Status::OK();
+}
+
+bool PreShapeNodeElimination::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger& logger) const {
+  if (!graph_utils::CanRemoveNode(graph, node, logger)) {
+    return false;
+  }
+
+  auto output_nodes = graph.GetConsumerNodes(node.OutputDefs()[0]->Name());
+
+  if (output_nodes.empty()) {
+    return false;
+  }
+
+  const Node* next_node = output_nodes[0];
```

Comment:
Should we check all output nodes?

Reply:
Should we just remove the cast if it only has one output?
or we can remove it even if it has more than one.
for example
         FastGelu
            |
          Cast
         /    \
      Shape   Shape
will be :

          FastGelu
        /         \
     Shape      Shape